src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -4515,7 +4592,17 @@ def _subroutine_handler(self, node, parent):
                 # attempt to recreate the prefix. We have to set shadowing to
                 # True as there is likely to be a RoutineSymbol for this
                 # function in any enclosing Container.
-
+                if not base_type:
+                    # The type of the return value was not specified in the
+                    # function prefix or in a local declaration and therefore
+                    # we have no explicit type information for it. Since we
+                    # default to adding `implicit none` when generating Fortran
+                    # we can't simply put this function into a CodeBlock as the
+                    # generated code won't compile.
+                    raise SymbolError(
```

Comment:
I am unsure if this should be a SymbolError (which for now where only raised from psyir/symbols )or an InternalError (which the parser sometimes do), but regardless what you decide, it needs to be in the docstring.

Reply:
I've decided to keep it as a SymbolError as InternalError implies a fault in PSyclone whereas this is a problem with the incoming code. I've also updated the error message a little. I've updated the docstring.